scripts/research/lib/research_harness.py

```
@@ -0,0 +1,579 @@
+def build_run_identity(
+    *,
+    hypothesis_id: str,
+    stage: str,
+    code_freeze: VerifiedCodeFreeze,
+    authorized_dataset: AuthorizedDataset,
+    command: Sequence[str],
+    seeds: Mapping[str, int] | None = None,
+) -> dict:
+    """Build deterministic provenance fields that must accompany a result."""
+    if not hypothesis_id or not stage:
+        raise ValueError("hypothesis_id and stage are required")
```

Ruff (0.16.2)

[warning] Avoid specifying long messages outside the exception class

(TRY003)